cli/lsp/testing.rs

```diff
@@ -0,0 +1,1959 @@
+      id: run.id,
+      stack: HashMap::new(),
+      tests,
+    }
+  }
+
+  fn add_step(&self, desc: &test::TestStepDescription) {
+    if let Ok(specifier) = ModuleSpecifier::parse(&desc.test.origin) {
+      let mut tests = self.tests.lock();
+      let entry =
+        tests
+          .entry(specifier.clone())
+          .or_insert_with(|| TestDefinitions {
+            discovered: Default::default(),
+            injected: Default::default(),
+            script_version: "1".to_string(),
+          });
+      let mut prev: lsp_custom::TestData = desc.into();
+      if let Some(stack) = self.stack.get(&desc.test.origin) {
+        for item in stack.iter().rev() {
+          let mut data: lsp_custom::TestData = item.into();
+          data.steps = Some(vec![prev]);
+          prev = data;
+        }
+        entry.injected.push(prev.clone());
+        let label = if let Some(root) = &self.maybe_root_uri {
+          specifier.as_str().replace(root.as_str(), "")
+        } else {
+          specifier
+            .path_segments()
+            .and_then(|s| s.last().map(|s| s.to_string()))
+            .unwrap_or_else(|| "<unknown>".to_string())
+        };
+        self
+          .channel
+          .send(TestingNotification::Module(
+            lsp_custom::TestModuleNotificationParams {
+              text_document: lsp::TextDocumentIdentifier { uri: specifier },
+              kind: lsp_custom::TestModuleNotificationKind::Insert,
+              label,
+              tests: vec![prev],
+            },
+          ))
+          .unwrap_or_else(|err| {
+            lsp_log!("{}", err);
+          });
+      }
+    }
+  }
+
+  /// Add a test which is being reported from the test runner but was not
+  /// statically identified
+  fn add_test(&self, desc: &test::TestDescription) {
+    if let Ok(specifier) = ModuleSpecifier::parse(&desc.origin) {
+      let mut tests = self.tests.lock();
+      let entry =
+        tests
+          .entry(specifier.clone())
+          .or_insert_with(|| TestDefinitions {
+            discovered: Default::default(),
+            injected: Default::default(),
+            script_version: "1".to_string(),
+          });
+      entry.injected.push(desc.into());
+      let label = if let Some(root) = &self.maybe_root_uri {
+        specifier.as_str().replace(root.as_str(), "")
+      } else {
+        specifier
+          .path_segments()
+          .and_then(|s| s.last().map(|s| s.to_string()))
+          .unwrap_or_else(|| "<unknown>".to_string())
+      };
+      self
+        .channel
+        .send(TestingNotification::Module(
+          lsp_custom::TestModuleNotificationParams {
+            text_document: lsp::TextDocumentIdentifier { uri: specifier },
+            kind: lsp_custom::TestModuleNotificationKind::Insert,
+            label,
+            tests: vec![desc.into()],
+          },
+        ))
+        .unwrap_or_else(|err| {
+          lsp_log!("{}", err);
+        });
```

Comment:
It would be less verbose if we got rid of the channel and instead created a function like the following:

```rs
fn send_testing_notification(&self, notification: TestingNotification) {
    let client = client.clone();
    tokio::task::spawn(async move {
      client.send_test_notification(notification).await;
    });
}
```

I kind of wonder if we should just do that internally in `client` though.

Reply:
Move it to the client method.